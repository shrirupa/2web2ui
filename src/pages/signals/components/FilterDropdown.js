import React, { useState, useCallback, useMemo } from 'react';
import {
  TextField,
  Popover,
  ActionList
} from '@sparkpost/matchbox';
import { ArrowDropDown } from '@sparkpost/matchbox-icons';

const FilterDropdown = ({ options, initialSelected, label, onChange }) => {
  const [selected, setSelected] = useState(initialSelected);

  const onItemClick = useCallback((e, value) => {
    const isSet = selected.indexOf(value) >= 0;
    setSelected(isSet ? [...selected.filter((item) => item !== value)] : [...selected, value]);
    e.stopPropagation();
  }, [selected]);

  const onClose = useCallback(() => {
    onChange(selected);
  }, [selected, onChange]);

  const formattedOptions = useMemo(
    () =>
      options.map(({ label, value }) => ({
        content: label,
        onClick: (e) => onItemClick(e, value),
        selected: selected.indexOf(value) >= 0
      })),
    [options, selected, onItemClick]
  );

  return (
    <div>
      <Popover
        trigger={<TextField readOnly value={label} prefix={selected.length ? `(${selected.length})` : null} suffix={<ArrowDropDown />} />}
        onClose={onClose}
      >
        <ActionList actions={formattedOptions} />
      </Popover>
    </div>
  );
};

export default FilterDropdown;
