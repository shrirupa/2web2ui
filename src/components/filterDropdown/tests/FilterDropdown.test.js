import { shallow } from 'enzyme';
import React from 'react';
import { FilterDropdown } from '../FilterDropdown';

describe('Component: FilterDropdown', () => {
  const props = {
    formName: 'myForm',
    label: 'My Label',
    id: 'my-filter-dropdown',
    namespace: 'status',
    displayValue: 'Status',
    change: jest.fn(),
    options: [
      { content: 'Active', name: 'active' },
      { content: 'Terminated', name: 'terminated' },
      { content: 'Suspended', name: 'suspended' },
    ],
  };

  const e = {
    stopPropagation: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<FilterDropdown {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders selected correctly', () => {
    wrapper.setProps({ values: { suspended: true } });
    expect(wrapper).toMatchSnapshot();
  });

  it('handles click', () => {
    wrapper.setProps({ values: { suspended: true } });
    const changeSpy = jest.spyOn(wrapper.instance().props, 'change');
    const eSpy = jest.spyOn(e, 'stopPropagation');
    wrapper.instance().handleActionClick(e, 'terminated');
    expect(changeSpy).toHaveBeenCalledWith(props.formName, `${props.namespace}.terminated`, true);
    expect(eSpy).toHaveBeenCalled();
  });
});
