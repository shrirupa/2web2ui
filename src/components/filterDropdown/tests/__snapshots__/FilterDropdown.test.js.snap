// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FilterDropdown renders correctly 1`] = `
<div>
  <Popover
    bottom={true}
    onClose={[Function]}
    right={true}
    trigger={
      <React.Fragment>
        <label
          className="hidden"
          htmlFor="filter-dropdown-my-filter-dropdown"
        >
          My Label
        </label>
        <TextField
          id="filter-dropdown-my-filter-dropdown"
          prefix={null}
          readOnly={true}
          resize="both"
          suffix={<ArrowDropDown />}
          type="text"
          value="Status"
        />
      </React.Fragment>
    }
  >
    <ActionList
      actions={
        Array [
          Object {
            "content": "Active",
            "name": "active",
            "onClick": [Function],
            "selected": undefined,
          },
          Object {
            "content": "Terminated",
            "name": "terminated",
            "onClick": [Function],
            "selected": undefined,
          },
          Object {
            "content": "Suspended",
            "name": "suspended",
            "onClick": [Function],
            "selected": undefined,
          },
        ]
      }
      groupByKey="section"
    />
  </Popover>
  <Field
    className="hidden"
    component="input"
    key="active"
    name="status.active"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
  <Field
    className="hidden"
    component="input"
    key="terminated"
    name="status.terminated"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
  <Field
    className="hidden"
    component="input"
    key="suspended"
    name="status.suspended"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
</div>
`;

exports[`Component: FilterDropdown renders selected correctly 1`] = `
<div>
  <Popover
    bottom={true}
    onClose={[Function]}
    right={true}
    trigger={
      <React.Fragment>
        <label
          className="hidden"
          htmlFor="filter-dropdown-my-filter-dropdown"
        >
          My Label
        </label>
        <TextField
          id="filter-dropdown-my-filter-dropdown"
          prefix="(1)"
          readOnly={true}
          resize="both"
          suffix={<ArrowDropDown />}
          type="text"
          value="Status"
        />
      </React.Fragment>
    }
  >
    <ActionList
      actions={
        Array [
          Object {
            "content": "Active",
            "name": "active",
            "onClick": [Function],
            "selected": false,
          },
          Object {
            "content": "Terminated",
            "name": "terminated",
            "onClick": [Function],
            "selected": false,
          },
          Object {
            "content": "Suspended",
            "name": "suspended",
            "onClick": [Function],
            "selected": true,
          },
        ]
      }
      groupByKey="section"
    />
  </Popover>
  <Field
    className="hidden"
    component="input"
    key="active"
    name="status.active"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
  <Field
    className="hidden"
    component="input"
    key="terminated"
    name="status.terminated"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
  <Field
    className="hidden"
    component="input"
    key="suspended"
    name="status.suspended"
    parse={[Function]}
    tabIndex="-1"
    type="checkbox"
  />
</div>
`;
